tooi/api/__init__.py

```diff
@@ -1,28 +1,57 @@
 import logging
 import time
 
-from httpx import Response
+from httpx import Response, RequestError
 from tooi.context import get_context
 from tooi.api.types import RequestParams
 from typing import Optional, Tuple, Unpack
 
 logger = logging.getLogger(__name__)
 
 
+class APIError(Exception):
+    """
+    Represents an error response from the API.
+    """
+
+    def __init__(self, message: str | None = None, cause: Exception | None = None):
+        assert (message or cause)
+        self.message = message or str(cause)
+        self.cause = cause
+        super().__init__(self.message)
+
+
+class ResponseError(Exception):
+    """Raised when the API returns a response with status code >= 400."""
+    def __init__(self, status_code: int, error: str | None, description: str | None):
+        self.status_code = status_code
+        self.error = error
+        self.description = description
+
+        msg = f"HTTP {status_code}"
+        msg += f". Error: {error}" if error else ""
+        msg += f". Description: {description}" if description else ""
+        super().__init__(msg)
+
+
 async def request(method: str, url: str, **kwargs: Unpack[RequestParams]) -> Response:
     ctx = get_context()
     start = time.time()
     logger.info(f"--> {method} {url}")
 
-    response = await ctx.auth.client.request(method, url, **kwargs)
+    try:
+        response = await ctx.auth.client.request(method, url, **kwargs)
+    except RequestError as exc:
+        raise (APIError(exc))
```

Comment:
This won't work since the first parameter to APIError constructor is the message.